bittide/src/Bittide/CaptureUgn.hs

```diff
@@ -0,0 +1,52 @@
+-- SPDX-FileCopyrightText: 2024 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+
+module Bittide.CaptureUgn (captureUgn) where
+
+import Clash.Explicit.Prelude hiding (PeriodToCycles)
+
+import Data.Maybe (fromMaybe, isJust)
+import Data.Tuple (swap)
+
+import Protocols
+import Protocols.Wishbone
+
+import Bittide.SharedTypes (Bytes)
+import Bittide.Wishbone (wbToVec)
+
+import qualified Clash.Prelude as C
+
+{- | Capture the UGN from the received data stream. The first `Just` value in the
+elastic buffer contains the rxCounter. Subtracting that from the txCounter results
+in the UGN. All subsequent values in the elastic buffer are bittide data.
+
+- Address 0: lsbs local counter
+- Address 1: msbs local counter
+- Address 2: lsbs remote counter
+- Address 3: msbs remote counter
+-}
+captureUgn ::
+  forall dom addrW.
+  ( C.HiddenClockResetEnable dom
+  , KnownNat addrW
+  ) =>
+  Signal dom (Unsigned 64) ->
+  Circuit
+    ( Wishbone dom 'Standard addrW (Bytes 4)
+    , CSignal dom (Maybe (BitVector 64))
+    )
+    (CSignal dom (Maybe (BitVector 64)))
```

Comment:
I actually expect that we can (and should) remove this `Maybe` since all communication within the network is scheduled.
You can simply apply `fromMaybe 0` to the output